[WIP] Initial SimpleSerialize (SSZ) spec

NatoliChris opened the pull request:
Initial specification outline for SSZ for #2.

* Basic information about serializing types. 
* Basic information about deserializing types.
* Checks to be performed for sanity. 
* Links to implementations.

TODO: 
* Ensure all implementations are listed.

---

There has been some discussion about `ssz` so if you notice any discrepancies would be great to have any input for changes :smile:. 

Also, if I've forgotten an implementation feel free to update the implementation list. There are some eth2.0 implementations that do not use `ssz` at this point in time (that I could see), but if I couldn't see it please make sure to add them.

Any help is appreciated, especially in terms of readability and making it easier to understand. We appreciate your time!

NatoliChris commented:
I think we should also add the test format once it has been finalised.

Thoughts? :smile:

hwwhww commented on specs/simpleserialize.md:
```diff
@@ -0,0 +1,230 @@
+
+## About
+
+`SimpleSerialize` was first proposed by Vitalik Buterin as the serializaiton
```
typo: `serializaiton` -> `serialization`

NatoliChris replied:
Good catch! Completely missed the misspelling :smile: Thanks!

hwwhww commented on specs/simpleserialize.md:
```diff
@@ -0,0 +1,230 @@
+
+| Constant       | Value | Definition                                                              |
+|:---------------|:-----:|:------------------------------------------------------------------------|
+| `LENGTH_BYTES` |   4   | Number of bytes used for the length added before the serialized object. |
```
`LENGTH_BYTES` should be applied in the followed description.

NatoliChris commented:
Thanks @hwwhww  :smile:

mratsim commented on specs/simpleserialize.md:
```diff
@@ -0,0 +1,230 @@
+ + [Serialize/Encode](#serializeencode)
+   - [uint: 8/16/24/32/64/256](#uint-816243264256)
+   - [Address](#address)
+   - [Hash32](#hash32)
```
With BLS we will have `Hash96` for the public keys maybe `Hash97` if it comes with a recovery bit.

NatoliChris replied:
@mratsim interesting point! I'll leave this open for discussion if we should tentatively add it in or leave it until we have more clarity on the point?

I'll draft it up and maybe add it into a PR?

mratsim replied:
a simple `Hash32 and other hash size` is probably enough at the moment

mratsim commented on specs/simpleserialize.md:
```diff
@@ -0,0 +1,230 @@
+
+#### Hash32
+
+The hash32 should already be a 32 byte length serialized data format. The safety
```
hash32 works for Keccak/Blake2b[0 ..< 32] not for BLS

NatoliChris replied:
Yep, great point! Did we address in point above?
 
I think we should have the `Hash97` and `Hash96` separately or address similar to the way `uint` is serialized/deserialized?
Thoughts?

mratsim replied:
I think like `uint`

mratsim commented on specs/simpleserialize.md:
```diff
@@ -0,0 +1,230 @@
+1. For each item in list:
+   1. serialize.
+   2. append to string.
+2. Get size of serialized string. Encode into a 4 byte integer.
```
This part is Python specific. This should read as

```Markdown
#### List/Vectors

1. Get the number of raw bytes to serialize: it is `len(list) * sizeof(element)`.
    Encode that as a 4-byte bigEndian uint32
2. Append your elements in a packed manner.

* Note on efficiency: consider using a container that does not need to iterate over all elements to get its length. For example Python lists, C++ vectors or Rust Vec.
```

Then your Python implementation notes.

The perf notes are to address the problem highlighted by @AlexeyAkhunov, if the container doe keep track of its length (for example LinkedLists) prefixing the length is inefficient.

NatoliChris replied:
Thanks! This is great. 

I think I'll add yours specifically, since it's more generic and should cater to larger of an audience! 

Thanks for pointing me to the notes, I'll make sure to run through them and give a reference! :smile:

mratsim replied:
You can refer to this post: https://github.com/ethereum/eth2.0-specs/issues/2#issuecomment-425626242

> ## On length prefix vs length suffix
>
> https://github.com/ethereum/beacon_chain/issues/94 and https://github.com/ethereum/eth2.0-pm/issues/8
>
> - The argument for moving at the end is because otherwise you would have to iterate on the data structure once to get the length then once again to get feed the data. ==> All languages even Haskell have a vector implementation that tracks length so you wouldn't need the first iteration to get length.
>
> - Furthermore putting length at the end means you need specific token to indicate end of data like ‘\0’ in cstring. This is causing a lot of issue in C code also this prevents reading by chunk into a buffer, you would have to read token by token (or read by chunk and backtrack).

NatoliChris replied:
Thanks @mratsim! :+1: 

I'll hold this PR open until we get more clarification/acceptance?

hwwhww replied:
I definitely don't want to block it, just thought it will be nice to have more eyes on it, like @mratsim. 👍👀 It looks good for me to merge it, as a clear document to be audited and to be disucussed.

NatoliChris commented:
Thanks @mratsim for the excellent comments! 
I'll provide the updates shortly! :smile:

djrtwo commented:
I'm going to give it a look this morning and merge. Just need to make sure some sort of "work in progress" signal is at the top of the doc

NatoliChris commented:
**Added Updates**

* Changed title to include explicit **WIP** status.
* Updated Hash types to include `hash96` and `hash97` for BLS.
    * *Note: It may be a point of discussion how this should be implemented. At this moment we went for simplicity and assumed it will be similar to the `uint` serialization technique.*
* Updated List to List/Vector and added @mratsim's suggestions.

There are currently 2 items that are to be discussed: 

## Container Serialization (@mratsim's comment):

> Nice work!
> One thing that is missing are "container types" serialization, for example ActiveState
> ```
> # ActiveState
> fields = {
>     # Attestations that have not yet been processed
>    'pending_attestations': [AttestationRecord],
>     # Most recent 2 * CYCLE_LENGTH block hashes, older to newer
>     'recent_block_hashes': ['hash32']
> }
> ```

I feel like this should be an open discussion around the appropriate methods of serializing a container.  @AlexeyAkhunov may also have input. 

## List/Vector Serialization

@mratsim's comment:

> The perf notes are to address the problem highlighted by @AlexeyAkhunov, if the container does keep > track of its length (for example LinkedLists) prefixing the length is inefficient.

> You can refer to this post: [#2 (comment)](https://github.com/ethereum/eth2.0-specs/issues/2#issuecomment-425626242)
> 
> > ## On length prefix vs length suffix
> > [ethereum/beacon_chain#94](https://github.com/ethereum/beacon_chain/issues/94) and [ethereum/eth2.0-pm#8](https://github.com/ethereum/eth2.0-pm/issues/8)
> > 
> > * The argument for moving at the end is because otherwise you would have to iterate on the data structure once to get the length then once again to get feed the data. ==> All languages even Haskell have a vector implementation that tracks length so you wouldn't need the first iteration to get length.
> > * Furthermore putting length at the end means you need specific token to indicate end of data like ‘\0’ in cstring. This is causing a lot of issue in C code also this prevents reading by chunk into a buffer, you would have to read token by token (or read by chunk and backtrack).


It might be appropriate like @hwwhww suggested to make this a separate discussion.

Thanks for all the input :smile:

djrtwo commented on specs/simpleserialize.md:
```diff
@@ -0,0 +1,302 @@
+| `byte_order` | Specifies [endianness:](https://en.wikipedia.org/wiki/Endianness) Big Endian or Little Endian. |
+| `len`        | Length/Number of Bytes.                                                                        |
+| `to_bytes`   | Convert to bytes. Should take parameters ``size`` and ``byte_order``.                          |
+| `from_bytes` | Convert form bytes to object. Should take ``bytes`` and ``byte_order``.                        |
```
Typo: "form" -> "from"

djrtwo commented on specs/simpleserialize.md:
```diff
@@ -0,0 +1,302 @@
+
+| Constant       | Value | Definition                                                              |
+|:---------------|:-----:|:------------------------------------------------------------------------|
+| `LENGTH_BYTES` |   4   | Number of bytes used for the length added before the serialized object. |
```
maybe say 'before a variable-length serialized object'. To be clear at this point that not all items need a length prefix

djrtwo commented on specs/simpleserialize.md:
````diff
@@ -0,0 +1,302 @@
+| Value is less than max           | ``2**int_size > value`` |
+
+```python
+buffer_size = int_size / 8
````
Some of the examples include the "Checks to perform" as `assert` statements, while others do not.

For consistency, I'd recommend including _all_ "check to perform" code snippets in the sample code

djrtwo commented on specs/simpleserialize.md:
````diff
@@ -0,0 +1,302 @@
+| Length of bytes can fit into 4 bytes | ``len(value) < 2**32`` |
+
+```python
+byte_length = (len(value)).to_bytes(LENGTH_BYTES, 'big')
````
include assertion

djrtwo commented on specs/simpleserialize.md:
```diff
@@ -0,0 +1,302 @@
+
+1. Get the number of raw bytes to serialize: it is `len(list) * sizeof(element)`.
+   * Encode that as a `4-byte` **big endian** `uint32`.
+2. Append your elements in a packed manner.
```
change "your" to "the"

djrtwo commented on specs/simpleserialize.md:
````diff
@@ -0,0 +1,302 @@
+**Example in Python**
+
+```python
+serialized_list_string = ''
````
For python bytes, use `b''`

djrtwo commented on specs/simpleserialize.md:
````diff
@@ -0,0 +1,302 @@
+```
+
+#### List/Vectors
+
````
Should add "Length of list can fit into 4 bytes" as a check in a "Check to perform" table as well as an assertion in the sample code

djrtwo commented on specs/simpleserialize.md:
```diff
@@ -0,0 +1,302 @@
+
+The decoding requires knowledge of the type of the item to be decoded. When
+performing decoding on an entire serialized string, it also requires knowledge
+of what order the objects have been serialized in.
```
"of the order in which the objects have been serialized"

djrtwo commented on specs/simpleserialize.md:
````diff
@@ -0,0 +1,302 @@
+
+```python
+bytes_length = int.from_bytes(rawbytes[current_index:current_index + LENGTH_BYTES], 'big')
+new_index = current_index + LENGTH_BYTES + bytes_lenth
````
typo: `bytes_lenth` -> `bytes_length`

djrtwo commented on specs/simpleserialize.md:
````diff
@@ -0,0 +1,302 @@
+```python
+bytes_length = int.from_bytes(rawbytes[current_index:current_index + LENGTH_BYTES], 'big')
+new_index = current_index + LENGTH_BYTES + bytes_lenth
+return rawbytes[current_index + LENGTH_BYTES:current_index+ LENGTH_BYTES +bytes_length], new_index
````
consider using two helper vars defined before the return statement for the indices -- `bytes_start` & `bytes_end`

djrtwo commented on specs/simpleserialize.md:
```diff
@@ -0,0 +1,302 @@
+entire length of the list.
+
+
+| Check type                          | code                                  |
```
Use "Check to perform" as in previous tables, and add the assertion in the sample code

NatoliChris commented:
Thanks @djrtwo! All applied :smile:. Thanks for the excellent comments!

djrtwo commented:
@NatoliChris Let's add a "TODO section/note" in the document for "container types" to make it clear it's missing.

After that, I'll merge this

NatoliChris commented:
Done :smile:
Thanks again @djrtwo !

djrtwo commented:
Excellent. Thanks!
